src/Hazelcast.Net/Clustering/Failover.cs

```diff
@@ -0,0 +1,164 @@
+        private Action<ClusterOptions> _clusterOptionsChanged;
+
+        internal Failover(ClusterState state, IHazelcastOptions options, ILoggerFactory loggerFactory)
+        {
+            if (options == null || options.Failover == null) throw new ArgumentNullException(nameof(options));
+            if (loggerFactory == null) throw new ArgumentNullException(nameof(loggerFactory));
+            if (state == null) throw new ArgumentNullException(nameof(state));
+
+            _logger = loggerFactory.CreateLogger<Failover>();
+            _state = state;
+            _isEnabled = options.Failover.Enabled;
+
+            if (!options.Failover.Clusters.Any() && options.Failover.Enabled)
+            {
+                throw new ConfigurationException("Failover is enabled but there is no cluster(s) provided. Please, check your configurations.");
+            }
+
+            //main cluster config is also an alternative cluster in a circular list (->blue->green->)
+            var clusterList = new List<ClusterOptions>
+            {
+                new ClusterOptions()
+                {
+                    Authentication = options.Authentication,
+                    ClusterName = options.ClusterName,
+                    Heartbeat = options.Heartbeat,
+                    LoadBalancer = options.LoadBalancer,
+                    Networking = options.Networking,
+                    WaitForConnectionMilliseconds = options.WaitForConnectionMilliseconds
+                }
+            };
+            clusterList.AddRange(options.Failover.Clusters);
+            _clusters = clusterList;
+
+            _clusterEnumerator = _clusters.GetEnumerator();
+            ResetToFirstCluster();
+
+            _isEnabled = options.Failover.Enabled;
+            _maxTryCount = options.Failover.TryCount;
+
+            HConsole.Configure(x => x.Configure<ClusterState>().SetPrefix("CLUST.FAILOVER"));
+        }
+
+        /// <summary>
+        /// Gets or sets action that will be executed after cluster options are switched
+        /// </summary>
+        public Action<ClusterOptions> ClusterOptionsChanged
+        {
+            get => _clusterOptionsChanged;
+            set
+            {
+                _state.ThrowIfPropertiesAreReadOnly();
+                _clusterOptionsChanged = value;
+            }
+        }
+
+        /// <summary>
+        /// Handles the event. If <see cref="CanSwitchClusterOptions"/> is true, and state is disconnected, it switches the <see cref="CurrentClusterOptions"/> to next one.
+        /// </summary>
+        public ValueTask OnClusterStateChanged(ClientState clientState)
+        {
+            //We have connected, reset the counter.
+            if (clientState == ClientState.Connected)
+            {
+                _currentTryCount = 0;
+            }
+            //we only count disconnected states
+            else if (clientState == ClientState.Disconnected && CanSwitchClusterOptions)
```

Comment:
Here I don't understand. if `clientState` is `Disconnected` it means that the client got disconnected... but shouldn't it first try to reconnect to the current cluster? maybe it was just a network glitch or something? is java doing the same, i.e. immediately look for another cluster without trying to reconnect? worse, we *are* trying to reconnect in the background anyways, I believe, so we may actually reconnect *while* you are switching. Unless I am missing something, that is a possibility ;)

Reply:
> ... but shouldn't it first try to reconnect to the current cluster? maybe it was just a network glitch or something?...

Could be but there is no exception to failover. Description is added to here at https://github.com/hazelcast/hazelcast-csharp-client/pull/587/commits/6f6c204b7a7e432c1258c9c932afa1999be3b78c as well as in the documentation.

Reply:
We do same now.